core/core-impl/src/main/java/ai/timefold/solver/core/impl/domain/variable/listener/support/violation/ListVariableTracker.java

```diff
@@ -0,0 +1,107 @@
+    private final List<Object> beforeVariableChangedEntityList;
+    private final List<Object> afterVariableChangedEntityList;
+
+    public ListVariableTracker(ListVariableDescriptor<Solution_> variableDescriptor) {
+        this.variableDescriptor = variableDescriptor;
+        beforeVariableChangedEntityList = new ArrayList<>();
+        afterVariableChangedEntityList = new ArrayList<>();
+    }
+
+    @Override
+    public VariableDescriptor<Solution_> getSourceVariableDescriptor() {
+        return variableDescriptor;
+    }
+
+    @Override
+    public void resetWorkingSolution(ScoreDirector<Solution_> scoreDirector) {
+        beforeVariableChangedEntityList.clear();
+        afterVariableChangedEntityList.clear();
+    }
+
+    @Override
+    public void beforeEntityAdded(ScoreDirector<Solution_> scoreDirector, Object entity) {
+
+    }
+
+    @Override
+    public void afterEntityAdded(ScoreDirector<Solution_> scoreDirector, Object entity) {
+
+    }
+
+    @Override
+    public void beforeEntityRemoved(ScoreDirector<Solution_> scoreDirector, Object entity) {
+
+    }
+
+    @Override
+    public void afterEntityRemoved(ScoreDirector<Solution_> scoreDirector, Object entity) {
+
+    }
+
+    @Override
+    public void afterListVariableElementUnassigned(ScoreDirector<Solution_> scoreDirector, Object element) {
+
+    }
+
+    @Override
+    public void beforeListVariableChanged(ScoreDirector<Solution_> scoreDirector, Object entity, int fromIndex, int toIndex) {
+        beforeVariableChangedEntityList.add(entity);
+    }
+
+    @Override
+    public void afterListVariableChanged(ScoreDirector<Solution_> scoreDirector, Object entity, int fromIndex, int toIndex) {
+        afterVariableChangedEntityList.add(entity);
+    }
+
+    public List<String> getEntitiesMissingBeforeAfterEvents(
+            List<VariableId<Solution_>> changedVariables) {
+        List<String> out = new ArrayList<>();
+        for (var changedVariable : changedVariables) {
+            if (!variableDescriptor.equals(changedVariable.variableDescriptor())) {
+                continue;
+            }
+            Object entity = changedVariable.entity();
+            if (!beforeVariableChangedEntityList.contains(entity)) {
+                out.add("Entity (" + entity
+                        + ") is missing a beforeListVariableChanged call for list variable ("
+                        + variableDescriptor.getVariableName() + ").");
+            }
+            if (!afterVariableChangedEntityList.contains(entity)) {
+                out.add("Entity (" + entity
+                        + ") is missing a afterListVariableChanged call for list variable ("
+                        + variableDescriptor.getVariableName() + ").");
+            }
+        }
+        beforeVariableChangedEntityList.clear();
+        afterVariableChangedEntityList.clear();
+        return out;
+    }
+
+    public TrackerDemand demand() {
+        return new TrackerDemand();
```

Comment:
No two demands for these classes will ever be equal. It a hack in order to get the variable listeners set up, since they need a Demand to get put into the logic.